### PR TITLE
merge next into master: changes that were lost in 2013, need to check whether they are still good.

### PR DESCRIPTION
**Pull request by ArneBab**

### CODE REVIEW

File: src/org/freenetproject/freemail/MailMessage.java [hunk 1]
```diff
@@ -576,4 +582,342 @@ public MailMessageHeader(String n, String v) {
 			this.val = v;
 		}
 	}
+
```
Comment (Bombe) [inline src/org/freenetproject/freemail/MailMessage.java:585]: Please use consistent indentation when adding large amounts of code.

File: src/org/freenetproject/freemail/MailMessage.java [hunk 2]
```diff
@@ -576,4 +582,342 @@ public MailMessageHeader(String n, String v) {
...
+			}
+		}
+
+		/**
+		 * Reads and decodes quoted-printable data until a canonical line has
+		 * been decoded.
+		 * @return a canonical line of message text
+		 * @throws IOException if an I/O error occurs while reading the input
+		 */
+		private String readQpLine() throws IOException {
```
Comment (Bombe) [inline src/org/freenetproject/freemail/MailMessage.java:648]: These methods should totally be broken out of here into their own classes, with their own tests and all.

File: src/org/freenetproject/freemail/transport/Channel.java [hunk 3]
```diff
@@ -434,17 +421,8 @@ boolean sendMessage(Bucket message, long messageId) throws ChannelTimedOutExcept
...
-		OutputStream messageOutputStream = null;
-		try {
-			messageOutputStream = fullMessage.getOutputStream();
-			BucketTools.copyTo(messageHeader, messageOutputStream, -1);
-			BucketTools.copyTo(message, messageOutputStream, -1);
-		} finally {
-			Closer.close(messageOutputStream);
-		}
+		
+		Bucket fullMessage = BucketChainBucket.constructReadOnly(new Bucket[] { messageHeader, message });
```
Comment (Olezha) [inline src/org/freenetproject/freemail/transport/Channel.java:425]: `freenet.support.io.BucketChainBucket` seems to have been removed.

File: src/org/freenetproject/freemail/ui/web/InboxToadlet.java [hunk 4]
```diff
@@ -348,16 +348,11 @@ private SortField(String name) {
...
-			if("subject".equals(field)) {
-				return SUBJECT;
-			}
-			if("from".equals(field)) {
-				return FROM;
-			}
-			if("date".equals(field)) {
-				return DATE;
+			try {
+				return SortField.fromString(field.toUpperCase());
```
Comment (Olezha) [inline src/org/freenetproject/freemail/ui/web/InboxToadlet.java:352]: Now this is recursion and StackOverflow. It meant `SortField.valueOf(field.toUpperCase())`

**Comment by Bombe** [comment]
This may have been superseded by https://github.com/hyphanet/plugin-Freemail/pull/43. The commit containing the `BucketChainBucket` and the commit introducing the stack overflow have been removed in that one. 😄

**Pull request merged**